File: kotlin/src/main/kotlin/com/bhengubv/circleai/companion/VoiceListener.kt
// VoiceListener.kt
//
// Kotlin port of CircleAI.Companion.IVoiceListener + VoiceCompanionListener —
// the C# reference (IVoiceListener.cs, VoiceCompanionListener.cs) is the EXACT
// spec. Bridges a voice pipeline with an ICompanionSession: transcribed
// utterances are forwarded to the session and the Companion's reply is surfaced
// back to subscribers.
//
// The C# type wires a CircleAI.Voice.VoicePipeline (via its .NET `Transcribed`
// event) to the session. That pipeline is not in the Kotlin tree, so — per the
// porting rules — the pipeline is an INJECTED dependency behind a minimal
// [IVoicePipeline] contract that exposes a cold Flow of transcriptions (the
// idiomatic Kotlin equivalent of the event). .NET events map to registerable
// listener callbacks. The forward-off-the-hot-path + swallow-failures semantics
// are preserved.

package com.bhengubv.circleai.companion

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import java.time.Instant
import java.util.concurrent.CopyOnWriteArrayList

// ---------------------------------------------------------------------------
// Event payloads
// ---------------------------------------------------------------------------

/**
 * Raised when a user utterance has been fully transcribed and forwarded to the
 * Companion session. Mirrors C# `UtteranceDetectedEventArgs`.
 */
data class UtteranceDetectedEvent(
    val text: String,
    val confidence: Float = 0f,
    val detectedAt: Instant = Instant.now(),
)

/**
 * Raised when the Companion has produced a reply to a voice utterance. Mirrors
 * C# `ResponseReadyEventArgs`.
 */
data class ResponseReadyEvent(
    val text: String,
    val originalUtterance: String,
    val completedAt: Instant = Instant.now(),
)

// ---------------------------------------------------------------------------
// Injected voice-pipeline port
// ---------------------------------------------------------------------------

/** One transcription result from the pipeline: [text] with a [confidence] and completion time. */
data class VoiceTranscription(val text: String, val confidence: Float, val completedAt: Instant = Instant.now())

/**
 * Minimal voice-pipeline port. The host's real pipeline (wake-word + ASR)
 * exposes its transcriptions as a cold [Flow]; [startAsync]/[stopAsync] control
 * capture. Injected so the neural ASR binding never leaks into this module.
 */
interface IVoicePipeline {
    fun transcriptions(): Flow<VoiceTranscription>
    suspend fun startAsync()
    suspend fun stopAsync()
    suspend fun closeAsync() {}
}

// ---------------------------------------------------------------------------
// IVoiceListener
// ---------------------------------------------------------------------------

/**
 * Bridges a voice pipeline with an [ICompanionSession]. Register listeners for
 * transcribed utterances and Companion replies; [startAsync]/[stopAsync] control
 * the underlying pipeline. Mirrors C# `IVoiceListener`.
 */
interface IVoiceListener {
    fun onUtteranceDetected(listener: (UtteranceDetectedEvent) -> Unit)
    fun onResponseReady(listener: (ResponseReadyEvent) -> Unit)
    suspend fun startAsync()
    suspend fun stopAsync()
    suspend fun closeAsync()
}

// ---------------------------------------------------------------------------
// VoiceCompanionListener
// ---------------------------------------------------------------------------

/**
 * Wires an [IVoicePipeline] to an [ICompanionSession]: each transcription is
 * forwarded to the session and the reply surfaced via [onResponseReady]. The
 * session call runs off the pipeline collection so it never blocks capture;
 * failures are logged, not thrown. Owns both the pipeline and the session —
 * [closeAsync] closes them.
 */
class VoiceCompanionListener(
    private val pipeline: IVoicePipeline,
    private val session: ICompanionSession,
) : IVoiceListener {

    private val utteranceListeners = CopyOnWriteArrayList<(UtteranceDetectedEvent) -> Unit>()
    private val responseListeners = CopyOnWriteArrayList<(ResponseReadyEvent) -> Unit>()

    @Volatile
    private var disposed = false

    private val scope = CoroutineScope(Dispatchers.Default + Job())
    private var collectJob: Job? = null

    override fun onUtteranceDetected(listener: (UtteranceDetectedEvent) -> Unit) {
        utteranceListeners.add(listener)
    }

    override fun onResponseReady(listener: (ResponseReadyEvent) -> Unit) {
        responseListeners.add(listener)
    }

    override suspend fun startAsync() {
        check(!disposed) { "VoiceCompanionListener is disposed" }
        if (collectJob == null) {
            collectJob = scope.launch {
                pipeline.transcriptions().collect { onTranscribed(it) }
            }
        }
        pipeline.startAsync()
    }

    override suspend fun stopAsync() {
        check(!disposed) { "VoiceCompanionListener is disposed" }
        pipeline.stopAsync()
    }

    /** Handle one transcription: notify subscribers, then forward to the session off the hot path. */
    private fun onTranscribed(t: VoiceTranscription) {
        if (disposed) return

        val detected = UtteranceDetectedEvent(
            text = t.text,
            confidence = t.confidence,
            detectedAt = t.completedAt,
        )
        for (l in utteranceListeners) l(detected)

        // Forward to the Companion asynchronously — never block the pipeline flow.
        scope.launch {
            try {
                val reply = session.sendAsync(t.text)
                if (!disposed) {
                    val ready = ResponseReadyEvent(
                        text = reply,
                        originalUtterance = t.text,
                        completedAt = Instant.now(),
                    )
                    for (l in responseListeners) l(ready)
                }
            } catch (ex: Exception) {
                System.err.println("VoiceCompanionListener: session failed for utterance '${t.text}': ${ex.message}")
            }
        }
    }

    override suspend fun closeAsync() {
        if (disposed) return
        disposed = true
        collectJob?.cancel()
        scope.cancel()
        pipeline.closeAsync()
        session.close()
    }
}

File: rust/src/companion/voice_listener.rs
//! voice_listener.rs
//!
//! `IVoiceListener` + `VoiceCompanionListener` — bridges the voice pipeline with
//! the Companion session. Ported from `IVoiceListener.cs` +
//! `VoiceCompanionListener.cs`: when the wake word fires and the user speaks, a
//! transcription is forwarded to the session and the Companion's reply is
//! surfaced via the `ResponseReady` event.
//!
//! The C# uses .NET events (multicast delegates) and a concrete `VoicePipeline`.
//! The Rust port models an event as a registered list of handler closures, the
//! pipeline as an injected [`IVoicePipeline`] seam whose transcriptions drive the
//! listener via [`VoiceCompanionListener::on_transcribed`], and the session as an
//! injected "send" closure (the C# calls `ICompanionSession.SendAsync`). This
//! keeps the event-bridge semantics — utterance-detected then response-ready,
//! never blocking the pipeline on failure — intact and testable.

use std::sync::{Arc, Mutex};

use chrono::{DateTime, Utc};

/// Arguments raised when a user utterance has been transcribed and forwarded.
#[derive(Debug, Clone, PartialEq)]
pub struct UtteranceDetectedEventArgs {
    /// Transcribed text of the user's utterance.
    pub text: String,
    /// Transcription confidence in `[0.0, 1.0]`.
    pub confidence: f32,
    /// UTC timestamp when the transcription completed.
    pub detected_at: DateTime<Utc>,
}

/// Arguments raised when the Companion has produced a reply to a voice utterance.
#[derive(Debug, Clone, PartialEq)]
pub struct ResponseReadyEventArgs {
    /// The Companion's reply text.
    pub text: String,
    /// The utterance that triggered this response.
    pub original_utterance: String,
    /// UTC timestamp when the Companion completed the reply.
    pub completed_at: DateTime<Utc>,
}

/// A transcription result produced by the pipeline.
#[derive(Debug, Clone, PartialEq)]
pub struct TranscriptionResult {
    pub text: String,
    pub confidence: f32,
}

/// The voice pipeline seam. A host wires the real wake-word + ASR pipeline; the
/// listener only needs start/stop and a way to be driven by transcriptions.
pub trait IVoicePipeline: Send + Sync {
    /// Begins listening for the wake word.
    fn start(&self);
    /// Stops listening.
    fn stop(&self);
}

/// The Companion "send" seam — forwards an utterance and returns the reply (or an
/// error, which the listener swallows like the C# try/catch). Mirrors
/// `ICompanionSession.SendAsync`.
pub type SessionSendFn = Arc<dyn Fn(&str) -> Result<String, String> + Send + Sync>;

type UtteranceHandler = Box<dyn Fn(&UtteranceDetectedEventArgs) + Send + Sync>;
type ResponseHandler = Box<dyn Fn(&ResponseReadyEventArgs) + Send + Sync>;

/// Bridges a voice pipeline with a Companion session.
pub trait IVoiceListener: Send + Sync {
    /// Registers a handler for the `UtteranceDetected` event.
    fn on_utterance_detected(&self, handler: UtteranceHandler);
    /// Registers a handler for the `ResponseReady` event.
    fn on_response_ready(&self, handler: ResponseHandler);
    /// Begins listening for the wake word.
    fn start(&self);
    /// Stops listening and cancels any in-flight activation.
    fn stop(&self);
}

/// Concrete [`IVoiceListener`] that wires a [`IVoicePipeline`] to a Companion
/// session-send closure. 1:1 with the C# `VoiceCompanionListener` behaviour.
pub struct VoiceCompanionListener {
    pipeline: Arc<dyn IVoicePipeline>,
    send: SessionSendFn,
    utterance_handlers: Arc<Mutex<Vec<UtteranceHandler>>>,
    response_handlers: Arc<Mutex<Vec<ResponseHandler>>>,
    disposed: Arc<Mutex<bool>>,
}

impl VoiceCompanionListener {
    /// Wires the pipeline to the session-send closure.
    pub fn new(pipeline: Arc<dyn IVoicePipeline>, send: SessionSendFn) -> Self {
        Self {
            pipeline,
            send,
            utterance_handlers: Arc::new(Mutex::new(Vec::new())),
            response_handlers: Arc::new(Mutex::new(Vec::new())),
            disposed: Arc::new(Mutex::new(false)),
        }
    }

    /// Drives one transcription through the bridge: raise `UtteranceDetected`,
    /// forward to the session, and (on success) raise `ResponseReady`. A no-op
    /// once disposed. Session failures are swallowed (the C# traces + continues),
    /// so a failing turn never raises `ResponseReady`.
    pub fn on_transcribed(&self, result: TranscriptionResult, completed_at: DateTime<Utc>) {
        if *self.disposed.lock().unwrap() {
            return;
        }

        let detected = UtteranceDetectedEventArgs {
            text: result.text.clone(),
            confidence: result.confidence,
            detected_at: completed_at,
        };
        for h in self.utterance_handlers.lock().unwrap().iter() {
            h(&detected);
        }

        match (self.send)(&result.text) {
            Ok(reply) => {
                if *self.disposed.lock().unwrap() {
                    return;
                }
                let ready = ResponseReadyEventArgs {
                    text: reply,
                    original_utterance: result.text.clone(),
                    completed_at: Utc::now(),
                };
                for h in self.response_handlers.lock().unwrap().iter() {
                    h(&ready);
                }
            }
            Err(_e) => {
                // Mirror the C# catch: trace + continue; never crash the pipeline.
            }
        }
    }

    /// Detaches the pipeline and stops raising events (mirrors `DisposeAsync`).
    pub fn dispose(&self) {
        let mut disposed = self.disposed.lock().unwrap();
        if *disposed {
            return;
        }
        *disposed = true;
        drop(disposed);
        self.pipeline.stop();
    }
}

impl IVoiceListener for VoiceCompanionListener {
    fn on_utterance_detected(&self, handler: UtteranceHandler) {
        self.utterance_handlers.lock().unwrap().push(handler);
    }

    fn on_response_ready(&self, handler: ResponseHandler) {
        self.response_handlers.lock().unwrap().push(handler);
    }

    fn start(&self) {
        if *self.disposed.lock().unwrap() {
            return;
        }
        self.pipeline.start();
    }

    fn stop(&self) {
        if *self.disposed.lock().unwrap() {
            return;
        }
        self.pipeline.stop();
    }
}
